tracing/tracing.go

```diff
@@ -0,0 +1,233 @@
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/prometheus/alertmanager/config"
+)
+
+const serviceName = "alertmanager"
+
+// Manager is capable of building, (re)installing and shutting down
+// the tracer provider.
+type Manager struct {
+	logger       log.Logger
+	done         chan struct{}
+	config       config.TracingConfig
+	shutdownFunc func() error
+}
+
+// NewManager creates a new tracing manager.
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger: logger,
+		done:   make(chan struct{}),
+	}
+}
+
+// Run starts the tracing manager. It registers the global text map propagator and error handler.
+// It is blocking.
+func (m *Manager) Run() {
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetErrorHandler(otelErrHandler(func(err error) {
+		level.Error(m.logger).Log("msg", "OpenTelemetry handler returned an error", "err", err)
+	}))
+	<-m.done
+}
+
+// ApplyConfig takes care of refreshing the tracing configuration by shutting down
+// the current tracer provider (if any is registered) and installing a new one.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	// Update only if a config change is detected. If TLS configuration is
+	// set, we have to restart the manager to make sure that new TLS
+	// certificates are picked up.
+	var blankTLSConfig config_util.TLSConfig
+	if reflect.DeepEqual(m.config, cfg.TracingConfig) && m.config.TLSConfig == blankTLSConfig {
+		return nil
+	}
+
+	if m.shutdownFunc != nil {
+		if err := m.shutdownFunc(); err != nil {
+			return fmt.Errorf("failed to shut down the tracer provider: %w", err)
+		}
+	}
+
+	// If no endpoint is set, assume tracing should be disabled.
+	if cfg.TracingConfig.Endpoint == "" {
+		m.config = cfg.TracingConfig
+		m.shutdownFunc = nil
+		otel.SetTracerProvider(noop.NewTracerProvider())
+		level.Info(m.logger).Log("msg", "Tracing provider uninstalled.")
+		return nil
+	}
+
+	tp, shutdownFunc, err := buildTracerProvider(context.Background(), cfg.TracingConfig)
+	if err != nil {
+		return fmt.Errorf("failed to install a new tracer provider: %w", err)
+	}
+
+	m.shutdownFunc = shutdownFunc
+	m.config = cfg.TracingConfig
+	otel.SetTracerProvider(tp)
+
+	level.Info(m.logger).Log("msg", "Successfully installed a new tracer provider.")
+	return nil
+}
+
+// Stop gracefully shuts down the tracer provider and stops the tracing manager.
+func (m *Manager) Stop() {
```

Comment:
There is a race condition between `ApplyConfig` and `Stop` around `m.shutdownFunc` as one goroutine can be updating the variable while another goroutine can be calling the variable. I think we either need a mutex or some docs to mention that these methods are not thread-safe.